File: leetcode/medium/238-product_of_array_except_self.ts
function productExceptSelf(nums: number[]): number[] {
    const n = nums.length;
    const left_right = new Array(n).fill(1);
    const right_left = new Array(n).fill(1);
    const res = new Array(n).fill(1);

    for (let i = 1; i < n; i++) {
        left_right[i] = left_right[i - 1] * nums[i - 1];
        right_left[n - i - 1] = right_left[n - i] * nums[n - i];
    }

    for (let i = 0; i < n; i++) {
        res[i] = left_right[i] * right_left[i];
    }

    return res;
};

File: leetcode/medium/238-product_of_array_except_self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> left(n);
        vector<int> right(n);

        int curr = 1;
        for (int i = 0; i < n; i++) {
            curr *= nums[i];
            left[i] = curr;
        }

        curr = 1;
        for (int i = n - 1; i >= 0; i--) {
            curr *= nums[i];
            right[i] = curr;
        }

        vector<int> res(n);
        for (int i = 0; i < n; i++) {
            if (i == 0) {
                res[i] = right[i+1];
            } else if (i == n - 1) {
                res[i] = left[i - 1];
            } else {
                res[i] = left[i - 1] * right[i + 1];
            }
        }

        return res;
    }
};
